pkg/merger/merger.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package merger
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+
+	"github.com/GoogleCloudPlatform/testgrid/config"
+	configpb "github.com/GoogleCloudPlatform/testgrid/pb/config"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+)
+
+type configList struct {
+	Target  string    `json:"Target"`
+	Sources []sources `json:"Sources"`
+}
+
+type sources struct {
+	Name     string `json:"Name"`
+	Location string `json:"Location"`
+}
+
+// ParseAndCheck parses and checks the configuration file for common errors
+func ParseAndCheck(data []byte) (target *gcs.Path, sources map[string]*gcs.Path, err error) {
+	cfg := &configList{}
+	err = yaml.Unmarshal(data, cfg)
+	if err != nil {
+		return
+	}
+
+	target, err = gcs.NewPath(cfg.Target)
+	if err != nil {
+		return
+	}
+
+	sources = make(map[string]*gcs.Path)
+	for _, source := range cfg.Sources {
+		if _, exists := sources[source.Name]; exists {
+			return nil, nil, fmt.Errorf("Duplicated name %s", source.Name)
+		}
+		path, err := gcs.NewPath(source.Location)
+		if err != nil {
+			return nil, nil, err
+		}
+		sources[source.Name] = path
+	}
+
+	if len(sources) == 0 {
+		return target, sources, errors.New("No shards to converge")
+	}
+	return
+}
+
+type mergeClient interface {
+	gcs.Opener
+	gcs.Uploader
+}
+
+// MergeAndUpdate gathers configurations from each path and merges them.
+// Puts the result at targetPath if confirm is true
+// Will skip an input config if it is invalid and skipValidate is false
+// Other problems are considered fatal and will return an error
+func MergeAndUpdate(ctx context.Context, client mergeClient, paths map[string]*gcs.Path, targetPath gcs.Path, skipValidate, confirm bool) error {
+	// Deserialize each proto
+	// TODO: Reading and validating can be done in parallel with a wait group.
+	// TODO: Cache the version for each source. Only read if they've changed.
+	shards := make(map[string]*configpb.Configuration)
+
+	for key, path := range paths {
+		if path == nil {
+			return fmt.Errorf("path at '%s' is nil", key)
+		}
+		cfg, err := config.ReadGCS(ctx, client, *path)
+		if err != nil {
+			return fmt.Errorf("Can't read config %s at %v: %w.", key, path, err)
```

Comment:
nit: Think these and other errors shouldn't be capitalized/punctuated

Reply:
Fixed